### PR TITLE
Add support for multi-arch manifest, with default arch on quay.io

### PR DESCRIPTION
**Pull request by deitch**
## Description

Currently, every supported arch (`amd64`, `arm64`, `ppc64le`) gets an image as `calico/node:<version>-<arch>`. In addition, the "archless" one `calico/node:<version>` is `amd64`.

This PR replaces the "archless" with a multi-arch manifest, so you can do `docker run calico/node:<version>` on any supported platform and it will pull down the correct one. 

This greatly simplifies using on non-`amd64` platforms, but also makes it easy to support Calico in multi-architecture kubernetes cluster, since the `DaemonSet` can just reference `calico/node:<version>` without needing to figure out which arch should run on which node.

It does this as follows:

* Add `push-manifests` and `push-non-manifests` to the `cd` target
* Create `push-manifests` target to push manifests for registries that support it
* Create `push-non-manifests` target to push default archless as `amd64` for registries that do not support multi-arch manifests
* Create `sub-single-tag-images-non-manifest-%` target to handle non-manifest registries having archless tags
* Remove the archless `amd64` tag and push from the existing targets so they can be handled cleanly by the above new targets

The manifests are pushed using the manifest-tool, which already is present in the go-build image.

As discussed with @tomdee and @fasaxc . cc @caseydavenport @mkumatag

This is the initial or "test" PR. Once this one is updated as needed, approved and merged in, I will open PRs for the other projects with distribution to match this.

### CODE REVIEW

**Comment by deitch** [comment]
> As you roll this out across the other repos, it would be great to try to keep the Makefiles as similar as possible across all the repos.

Sure thing. Can you update the docs that describe our `Makefile` standards so I have a doc (not just this template) to follow? I think the targets are [here](https://github.com/projectcalico/calico/blob/master/BUILDING_CALICO.md#build-targets), so probably the right place.

I will update based on your recommended changes.

File: Makefile [hunk 1]
```diff
@@ -66,11 +73,16 @@ BUILD_IMAGE?=calico/node
 PUSH_IMAGES?=$(BUILD_IMAGE) quay.io/calico/node
 RELEASE_IMAGES?=gcr.io/projectcalico-org/node eu.gcr.io/projectcalico-org/node asia.gcr.io/projectcalico.org/node us.gcr.io/projectcalico.org/node
 
-ifeq ($(RELEASE),true)
 # If this is a release, also tag and push additional images.
+ifeq ($(RELEASE),true)
 PUSH_IMAGES+=$(RELEASE_IMAGES)
 endif
 
+# remove frin the list to push to manifest any registries that do not support multi-arch
```
[the remark this answers is not in the document]
Comment (deitch) [reply Makefile]: Oops. And I cannot blame this one on the wacky MacBook Pro keyboards. :-)

File: Makefile [hunk 2]
```diff
@@ -238,20 +243,33 @@ push: imagetag $(addprefix sub-single-push-,$(call escapefs,$(PUSH_IMAGES)))
...
 sub-push-%:
 	$(MAKE) push ARCH=$* IMAGETAG=$(IMAGETAG)
 
-## tag images of one arch
-tag-images: imagetag $(addprefix sub-single-tag-images-,$(call escapefs,$(PUSH_IMAGES)))
+## push multi-arch manifest where supported
+push-manifests: imagetag  $(addprefix sub-manifest-,$(call escapefs,$(PUSH_MANIFEST_IMAGES)))
+sub-manifest-%:
+	# Docker login to hub.docker.com required before running this target as we are using $(HOME)/.docker/config.json holds the docker login credentials
+	docker run -t --entrypoint /bin/sh -v $(HOME)/.docker/config.json:/root/.docker/config.json $(CALICO_BUILD) -c "/usr/bin/manifest-tool push from-args --platforms $(call join_platforms,$(VALIDARCHES)) --template $(call unescapefs,$*:$(IMAGETAG))-ARCH --target $(call unescapefs,$*:$(IMAGETAG))"
```
[the remark this answers is not in the document]
Comment (deitch) [reply Makefile]: I cannot think of a better way either, but I like the idea of defining it as a variable.

```makefile
DOCKER_CONFIG ?= $(HOME)/.docker/config
# and then
sub-manifest-%:
	docker run ... $(DOCKER_CONFIG):/root/.docker/config.json ...
```

> Is there a Docker doc which defines the location of this file?

Yep, will put in comment.

**Comment by deitch** [comment]
BTW, there is _so_ much boilerplate we have to do in our makefiles. I wish `Makefile` had an "include by URL", but they only have by path. Still, I do have some ideas.

Let's get these multi-arch merged in, and then we can discuss.

File: Makefile (hunk 2 above)
[the remark this answers is not in the document]
Comment (deitch) [reply Makefile]: Well, not true, I _can_ think of a better way: use `docker manifest`. But it isn't ready yet. When it is, we can replace all of this. Until then, we live with it as is, so temporary ugliness.

**Comment by deitch** [comment]
Updated per your suggestions.

**Comment by deitch** [comment]
@tomdee the CI failed with `make ci` which in turn failed from `make st`. Can you see if you understand why? Failed build is [here](https://semaphoreci.com/calico/node/branches/pull-request-42/builds/1)

**Comment by deitch** [comment]
Ignore my previous comment, as CI passed. Must have been a timing issue. All looking good now! :-)

**Comment by mkumatag** [comment]
I'm very curious to see this PR get into the release. Sorry I couldn't find much time to focus on these PR's nowadays.. I really appreciate you guys.. kudos @deitch @tomdee @fasaxc Great work... 🎉

**Pull request merged**